File: leet_BTREE_0671_second_min.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findSecondMinimumValue(self, root: Optional[TreeNode]) -> int:
        return [
            # self.BFS_with_set,
            # self.BFS_economic,
            self.DFS,
        ][0](root)

    def DFS(self, root: Optional[TreeNode]) -> int:
        res0, res1 = 2147483648, 2147483648
        def recurse(node: Optional[TreeNode]) -> None:
            if not node:
                return
            n = node.val
            nonlocal res0, res1
            if n < res0:
                res1 = res0
                res0 = n
            elif n != res0 and n < res1:
                res1 = n
            recurse(node.left)
            recurse(node.right)
        recurse(root)
        if res1 != 2147483648:
            return res1
        return -1

    def BFS_economic(self, root: Optional[TreeNode]) -> int:

        res = [2147483648, 2147483648]
        Q = deque()
        Q.append(root)
        while Q:
            node = Q.popleft()
            n = node.val
            if n < res[0]:
                res[1] = res[0]
                res[0] = n
            elif n != res[0] and n < res[1]:
                res[1] = n
            if node.left:
                Q.append(node.left)
            if node.right:
                Q.append(node.right)
        if res[1] == 2147483648:
            return -1
        return res[1]

    def BFS_with_set(self, root: Optional[TreeNode]) -> int:

        Q = deque()
        Q.append(root)
        S = set()
        while Q:
            node = Q.popleft()
            n = node.val
            S.add(n)
            if node.left:
                Q.append(node.left)
            if node.right:
                Q.append(node.right)
        res = list(S)
        if len(res) < 2:
            return -1
        return sorted(res)[1]

File: leet_BTREE_0671_second_min.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

long long res0 = 2147483648;
long long res1 = 2147483648;

long long res = 2147483648;

void DFS_1(struct TreeNode *);
void DFS_2(struct TreeNode *, int *);

int findSecondMinimumValue(struct TreeNode* root) {

    // vers. 1
    res0 = 2147483648;
    res1 = 2147483648;
    DFS_1(root);
    // return res1 ^ 2147483648 ? (int) res1 : -1;

    /* ****** ****** ****** */

    // vers. 2
    res = 2147483648;
    DFS_2(root, &(root->val));
    return res ^ 2147483648 ? (int) res : -1;
}

void DFS_2(struct TreeNode * node, int * curr) {

    if (!node)
        return ;
    int n = node->val;
    if ( n != *curr && n < res)
        res = n;
    if (res <= n)
        return ;
    DFS_2(node->left, curr);
    DFS_2(node->right, curr);
}

void DFS_1(struct TreeNode * node) {

    if (!node)
        return ;
    int n = node->val;
    if (n < res0)
    {
        res1 = res0;
        res0 = n;
    }
    else if (n != res0 && (n < res1))
    {
        res1 = n;
    }
    DFS_1(node->left);
    DFS_1(node->right);
}

